### PR TITLE
add support for 0x80 SUBACK return code

### LINKED ISSUE #72: SUBACK not returned when subscriptions are denied.
I came accross this while developing a custom ACL plugin based on `vmq_pgauth_demo`, but this behaviour happens on a vanilla VerneMQ deployment as well (as of `0.12.2`) with `vmq_acl`.

If a client issues a `SUBSCRIBE` and it is denied by some `auth_on_subscribe` plugin, no `SUBACK` packet is returned to the client (leaving the client without any feedback and blocking synchronous clients).

@ioolkos @dergraf could you reproduce this behaviour? Is there any workaround?

### PR DESCRIPTION
**Pull request by dergraf**
this commit is required to eventually fix erlio/vernemq#72
@ioolkos @larshesel please review

### CODE REVIEW

**Comment by dergraf** [comment]
just stumbled over a merge mistake.. stay tuned

**Comment by dergraf** [comment]
a few merge conflicts and a git squash later... please review @ioolkos cc: @larshesel 

the commit for the vmq_server part is ready, but I wait with merging until this one is closed

**Comment by dergraf** [comment]
@larshesel please review and merge cc:@ioolkos

**Comment by larshesel** [comment]
@dergraf I'm on it

File: src/vmq_parser.erl [hunk 1]
```diff
@@ -401,11 +407,16 @@ gen_pubrel(MId) ->
 gen_pubcomp(MId) ->
     iolist_to_binary(serialise(#mqtt_pubcomp{message_id=MId})).
 
-gen_subscribe(MId, Topic, Qos) ->
-    iolist_to_binary(serialise(#mqtt_subscribe{topics=[{ensure_binary(Topic), Qos}], message_id=MId})).
```
Comment (larshesel) [inline src/vmq_parser.erl:405]: Why don't we have to 'ensure_binary' on the topic anymore?

Ignore that - read the code wrong - it's still there

File: src/vmq_parser.erl [hunk 2]
```diff
@@ -311,17 +314,20 @@ serialise_len(N) when N =< ?LOWBITS ->
...
 
-serialise_topics(_, [], Topics) ->
-    Topics;
 serialise_topics(?SUBSCRIBE = Sub, [{Topic, QoS}|Rest], Acc) ->
     serialise_topics(Sub, Rest, [utf8(vmq_topic:unword(Topic)), <<0:6, QoS:2>>|Acc]);
 serialise_topics(?UNSUBSCRIBE = Sub, [Topic|Rest], Acc) ->
-    serialise_topics(Sub, Rest, [utf8(vmq_topic:unword(Topic))|Acc]).
+    serialise_topics(Sub, Rest, [utf8(vmq_topic:unword(Topic))|Acc]);
+serialise_topics(_, [], Topics) ->
+    Topics.
```
Comment (larshesel) [inline src/vmq_parser.erl]: What is this special case used for?

**Comment by larshesel** [comment]
It looks good to me - the only thing that was a bit confusing is the name of the compare_frame test function - it actually tests that parse/serialize are inverses of each other - that's not easy to understand from the name. Maybe call it something like `parse_serialize_test` would be a better name.

**Comment by dergraf** [comment]
@larshesel just figured out that my code had a major bug.

**Comment by larshesel** [comment]
@dergraf oh - what was the problem?

**Comment by dergraf** [comment]
```
<<0:6, 128:2>> gives <<0>> 
```

**Pull request merged**